boltons/pathutils.py

```diff
@@ -0,0 +1,212 @@
+        foo_new.tar.gz
+    """
+    # Breakup path
+    orig_dpath, fname = split(path)
+    if multidot:
+        # The first dot defines the extension
+        parts = fname.split('.', 1)
+        orig_base = parts[0]
+        orig_ext = '' if len(parts) == 1 else '.' + parts[1]
+    else:
+        # The last dot defines the extension
+        orig_base, orig_ext = splitext(fname)
+    # Replace parts with specified augmentations
+    if dpath is None:
+        dpath = orig_dpath
+    if ext is None:
+        ext = orig_ext
+    if base is None:
+        base = orig_base
+    # Recombine into new path
+    new_fname = ''.join((prefix, base, suffix, ext))
+    newpath = join(dpath, new_fname)
+    return newpath
+
+
+def shrinkuser(path, home='~'):
+    """
+    Inverse of ````os.path.expanduser````
+
+    Args:
+        path (PathLike): path in system file structure
+        home (str): symbol used to replace the home path. Defaults to '~', but
+            you might want to use '$HOME' or '%USERPROFILE%' instead.
+
+    Returns:
+        PathLike: path: shortened path replacing the home directory with a tilde
+
+    Example:
+        >>> path = expanduser('~')
+        >>> assert path != '~'
+        >>> assert shrinkuser(path) == '~'
+        >>> assert shrinkuser(path + '1') == path + '1'
+        >>> assert shrinkuser(path + '/1') == join('~', '1')
+        >>> assert shrinkuser(path + '/1', '$HOME') == join('$HOME', '1')
+    """
+    path = normpath(path)
+    userhome_dpath = userhome()
+    if path.startswith(userhome_dpath):
+        if len(path) == len(userhome_dpath):
+            path = home
+        elif path[len(userhome_dpath)] == os.path.sep:
+            path = home + path[len(userhome_dpath):]
+    return path
+
+
+def expandpath(path):
+    """
+    Wrapper around expanduser and expandvars.
+
+    Less aggressive than truepath. Only expands environs and tilde. Does not
+    change relative paths to absolute paths.
+
+    Args:
+        path (PathLike): string representation of a path
+
+    Returns:
+        PathLike : expanded path
+
+    Example:
+        >>> assert normpath(expandpath('~/foo')) == join(userhome(), 'foo')
+        >>> assert expandpath('foo') == 'foo'
+    """
+    path = expanduser(path)
+    path = expandvars(path)
+    return path
+
+
+def userhome(username=None):
+    """
+    Returns the user's home directory.
+
+    Args:
+        username (str, default=None): name of a user on the system. If not
+            specified, the current user is inferred.
+
+    Returns:
+        PathLike: userhome_dpath: path to the home directory
+
+    Example:
+        >>> import getpass
+        >>> username = getpass.getuser()
+        >>> assert userhome() == expanduser('~')
+        >>> assert userhome(username) == expanduser('~')
+    """
+    if username is None:
+        # get home directory for the current user
+        if 'HOME' in os.environ:
+            userhome_dpath = os.environ['HOME']
```

Comment:
Small thing, but I think I prefer this to be structured as a guard statement and just early return, so the big else block can be dedented.

Also, why not do `expanduser('~')` here?

Reply:
I tend to prefer code that doesn't have early return statements. It makes it much easier to copy-paste segments into IPython when I'm doing interactive development. I also like having a single return as the last line of a function, so you know that whatever code path is taken must always populate that specific return variable. However, this is only a minor preference, and I'm willing to make this modification. 

However, I would prefer to keep the existing implementation without using `expanduser('~')`. The expanduser function is actually defined in different files in the stdlib depending on your OS. I think its nicer to have a single code path that shows behavior on all system. Also this code cuts out a lot of the extra work that `expanduser('~')` would have to do because we aren't replacing anything in a string we are only finding the path to the user's home directory. While using expanduser is more concise I think the current implementation demystifies what expanduser is actually doing under the hood. My vote is to keep this, but I'll yield to your opinion if you feel strongly about it.